File: 0308-range-sum-query-2d-mutable/0308-range-sum-query-2d-mutable.java
class NumMatrix {

    private final int rows;
    private final int cols;
    private final int[][] matrix;
    private final int[][] bit;

    public NumMatrix(int[][] matrix) {
        rows = matrix.length;
        cols = rows == 0 ? 0 : matrix[0].length;

        this.matrix = new int[rows][cols];
        this.bit = new int[rows + 1][cols + 1];

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                update(row, col, matrix[row][col]);
            }
        }
    }

    public void update(int row, int col, int val) {
        int difference = val - matrix[row][col];
        matrix[row][col] = val;

        for (int i = row + 1; i <= rows; i += i & -i) {
            for (int j = col + 1; j <= cols; j += j & -j) {
                bit[i][j] += difference;
            }
        }
    }

    public int sumRegion(int row1, int col1, int row2, int col2) {
        return prefixSum(row2, col2)
                - prefixSum(row1 - 1, col2)
                - prefixSum(row2, col1 - 1)
                + prefixSum(row1 - 1, col1 - 1);
    }

    private int prefixSum(int row, int col) {
        int sum = 0;

        for (int i = row + 1; i > 0; i -= i & -i) {
            for (int j = col + 1; j > 0; j -= j & -j) {
                sum += bit[i][j];
            }
        }

        return sum;
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * obj.update(row,col,val);
 * int param_2 = obj.sumRegion(row1,col1,row2,col2);
 */

File: premium questions/0308-range-sum-query-2d-mutable/0308-range-sum-query-2d-mutable.cpp
#define rep(i, a, b) for(int i=(a); i<(b); ++i)
#define sz(x) (int)(x).size()
typedef vector<int> vi;
typedef vector<vector<int>> vii;

class Bit {
public:
    int R;
    int C;
    vii vii_;
    Bit(int m, int n): R(m), C(n), vii_(m, vi(n)) {}
    void update(int row, int col, int dif) {
        int r = row;
        while (r < R) {
            int c = col;
            while (c < C) {
                vii_[r][c] += dif;
                c |= c+1;
            }
            r |= r+1;
        }
    }
    int query(int row, int col) {
        int sm = 0;
        int r = row;
        while (r) {
            int c = col;
            while (c) {
                sm += vii_[r-1][c-1];
                c &= c-1;
            }
            r &= r-1;
        }
        return sm;
    }
};

class NumMatrix {
public:
    Bit b;
    vii& mat;
    NumMatrix(vector<vector<int>>& mat): b(sz(mat), sz(mat[0])), mat(mat) {
        rep(r, 0, sz(mat)) {
            rep(c, 0, sz(mat[0])) {
                b.update(r, c, mat[r][c]);
            }
        }
    }
    
    void update(int row, int col, int val) {
        int dif = val - mat[row][col];
        mat[row][col] = val;
        b.update(row, col, dif);
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        // cout << b.query(0, 0) << endl;
        // cout << b.query(1, 1) << endl;
        // cout << b.query(1, 2) << endl;
        // cout << b.query(2, 1) << endl;
        // cout << b.query(2, 2) << endl;
        // return 0;
        int sm1 = b.query(row1, col1);
        int sm2 = b.query(row1, col2+1);
        int sm3 = b.query(row2+1, col1);
        int sm4 = b.query(row2+1, col2+1);
        return (sm1 + sm4) - (sm2 + sm3);
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * obj->update(row,col,val);
 * int param_2 = obj->sumRegion(row1,col1,row2,col2);
 */